### PR TITLE
Make input Options case sensitive and error on unused Options

### PR DESCRIPTION
**Pull request by ZedThree**
As discussed in #2209, this makes `Options` case sensitive, which is a breaking change.

It also introduces another breaking change: by default, any unused options after an initial call to `rhs` is an error. This can be toggled with `input:error_on_unused_options` (I used this instead of `input:strict` as it's more explicit: `strict` sounds like turning it off might allow case-insensitive options).

I had to fix some other bits for this to work: some `Solver` unit tests were a little fragile, and constructing `Options` from a braced init list turned out to be broken in a fun way. Writing `Options::getUnused` also turned out to be way more difficult than I was expecting. `Options` is actually a little more difficult to use than I thought. 

I'm not overly happy with this check being done in `Solver::solve` -- it's the obvious place, but it doesn't really feel like `Solver` 's responsibility. I want to also add `input:help` which prints out all the seen options, and again, `Solver::solve` is the obvious place, but it feels like it should live somewhere "higher up".

Also to be done, add Levenshtein distance checking to point out likely typos

Also, I've added `toString(Options)` which ignores the doc strings and source. It's very close to just completely replacing `OptionsINI`, so I'd also like to work out how to merge all that.

I figure all these changes are very likely to break some tests, so I'm looking forward to seeing what needs fixing!

### CODE REVIEW

File: include/options.hxx
Comment (bendudson) [inline include/options.hxx:670]: Is it worth having a shortcut here to check if they are the same object `(&lhs == &rhs)`?

File: src/sys/options.cxx
Comment (bendudson) [inline src/sys/options.cxx:69]: Is it worth instead making full_name a function, which iterates through its parents? Of course that raises the problem of setting pointers to parents... Another possibility is to not store the full name or parent, but that then puts it on wherever the full name is needed to work it out.

File: src/sys/options.cxx
Comment (bendudson) [inline src/sys/options.cxx:78]: Sorry! I missed this before. Maybe worth checking where full_name is actually needed, and if it could be worked out when needed e.g. as the tree of options is traversed.

File: src/sys/options.cxx
Comment (bendudson) [inline src/sys/options.cxx:702]: Why the two-step thing here, and not just have `std::string toString(const Options& value)`? Is there another place where `void toString(const Options& value, std::string& fout)` is useful?

File: src/sys/options.cxx
Comment (ZedThree) [reply src/sys/options.cxx:78]: I needed it for printing out subsections correctly, i.e.

```
[section:subsection]
value = thing
```

but, yes, maybe this could be done when traversing the tree

File: src/sys/options.cxx
Comment (ZedThree) [reply src/sys/options.cxx:702]: Probably not, I copied the implementation from the python version of `BoutOptions` in `data.py` -- I had a vague memory that the two-step-ness was needed there, but this took so long to get working, I didn't want to spent more time refining it already. Will have another look when I come back to this!

**Comment by ZedThree** [comment]
As predicted, lots of test failures. It does look like a lot, if not all, are due to the case sensitivity:

```
The following tests FAILED:
	  5 - test-communications (Failed)
	  6 - test-coordinates-initialization (Failed)
	  8 - test-delp2 (Failed)
	  9 - test-drift-instability (Failed)
	 10 - test-drift-instability-staggered (Failed)
	 11 - test-fieldgroupComm (Failed)
	 12 - test-griddata (Failed)
	 16 - test-interchange-instability (Failed)
	 18 - test-interpolate-z (Failed)
	 19 - test-invertable-operator (Failed)
	 20 - test-invpar (Failed)
	 24 - test-laplace-petsc3d (Failed)
	 28 - test-naulin-laplace (Failed)
	 29 - test-petsc-laplace (Failed)
	 30 - test-petsc-laplace-MAST-grid (Failed)
	 31 - test-restart-io (Failed)
	 32 - test-restart-io-hdf5 (Failed)
	 33 - test-restarting (Failed)
	 34 - test-slepc-solver (Failed)
	 37 - test-solver (Failed)
	 38 - test-squash (Failed)
	 40 - test-stopCheck-file (Failed)
	 41 - test-twistshift (Failed)
	 42 - test-twistshift-staggered (Failed)
	 43 - test-vec (Failed)
	 44 - test-yupdown (Failed)
	 45 - test-yupdown-weights (Failed)
```

**Comment by ZedThree** [comment]
I've started looking through the failures. If we're going case-sensitive, we're going to need to decide on whether we want to keep existing spellings/cases or change them to match common usage. For example, the `NOUT` option is currently all caps, but very split in the tests and examples whether it is all caps or not. On the other hand, the `mesh:dd*:First` option is capital `F` but _all_ the input files spell it `first`.

If we provide a tool for correcting case in input files, this should have minimal impact on users so I'm inclined to make most/all options in the library lowercase.

**Comment by ZedThree** [comment]
Still WIP, and expecting tests to fail, as I've not yet fixed the input names.

TODO:
- fix input options to prefer lowercase by default (maybe except where the variable itself is ALL CAPS? but maybe also fix that too?)
- add fixes for input options to `bin/bout-v5-input-upgrader.py` and apply to all input files
- remove other unused options from suggested alternatives
- remove unused options with a default source from the set of unused options (we don't want to error because `run:version = v5.0.0-alpha` wasn't used -- this is really an "output" option, not an input) 
- add `input:help` to display inputs and docstrings
- add custom fmt formatter for `Options` to enable printing either as INI-style or as flattened keys, with/without docstrings/type/source etc?

File: include/options.hxx
Comment (ZedThree) [inline include/options.hxx:289]: This might actually be nicer as the `Options` itself (or a copy), rather than just the name

File: src/solver/solver.cxx
Comment (ZedThree) [inline src/solver/solver.cxx:513]: Maybe this whole chunk should be broken out into a function somewhere else? I'm still not keen on this living in `Solver::solve`, but moving it out requires something like `BOUTMAIN` calling `PhysicsModel::init/run_rhs` to get all the options.

Also the `input:help` functionality needs to essentially be called in the same place, and then we may have to be careful that nothing has tried to write to any files, or we'll end up clobbering stuff. I was wondering if we could somehow run in a "fake" location somewhere to avoid this. We'd still need access to the input file and grid file, etc.

File: src/sys/options.cxx
Comment (ZedThree) [inline src/sys/options.cxx:141]: There are various choices we could make about how this actually functions. I've chosen a default max distance of 4, so an option with more than 4 edits from the search string aren't counted, for example: `some_string` won't match `another_string` (by default).

It also matches an "unqualified name" (e.g. `first`) against a "fully qualified" name (e.g. `mesh:ddx:first`) with a distance of 1. But it won't try to strip the parent sections and match that. So `mesh:ddx:first` won't match `some:other:first`.

Another check is matching different cases: `THIS` matches `this` matches `ThIs`, all distance 1.

These choices were a little arbitrary, so definitely happy to hear suggestions on other choices

**Comment by ZedThree** [comment]
Might need to turn off clang-tidy-review on all the googletest stuff...

File: src/solver/solver.cxx
Comment (bendudson) [reply src/solver/solver.cxx:513]: It might be nice to have the init and first RHS call be done in a separate call, rather than inside `Solver`. I remember a long time ago (2013?) we had to change how Solver initialised and ran, because of the weird interface PETSc had then. Perhaps its worth revisiting?

**Comment by ZedThree** [comment]
Failing on unused options has a "fun" case where we have something like this:

```cpp
    if (!estatic && !ZeroElMass) {
      apar_solver = Laplacian::create(globalOptions->getSection("aparsolver"));
    }
```

Here, everything in the `aparsolver` section is unused. I can see two ways of dealing with this:

1. just comment out the section in the input file
2. unconditionally create `apar_solver`

More generically, do we expect the user to comment out or otherwise remove inputs when some other option means they go unused? Or to restructure their code to always read the options at least? Or perhaps to just confirm they've spelt everything correctly and then set `input:error_on_unused=false`?

**Comment by bendudson** [comment]
I think an error message suggesting that the user checks, and then sets `input:error_on_unused=false` would be fine. We could perhaps have a function which says that an option is not currently used, but may be e.g.
```
if (!estatic && !ZeroElMass) {
   apar_solver = Laplacian::create(globalOptions->getSection("aparsolver"));
} else {
   globalOptions->getSection("aparsolver").setConditionallyUnused();
}
```
but with a better name. Perhaps an attribute in the `Option` so that the check for unused options skips that option or section.

**Comment by ZedThree** [comment]
Almost there... I need to fix `boutdata.data.BoutOptionsFile`, as instead of renaming e.g. the section "All" to "all", it instead just completely removes it. Which makes the input file upgrader a little dangerous.

Error messages should now be a bit more useful. Here's an example of blob2d:

```

There were unused input options:
-----
NOUT = 50
TIMESTEP = 50

[All]
bndry_all = neumann
scale = 0.0

[phiBoussinesq]
flags = 0

[phiSolver]
flags = 0
fourth_order = true
pctype = user
type = petsc

[phiSolver:precon]
filter = 0.
flags = 49152

-----
It's possible you've mistyped some options. BOUT++ input arguments are
now case-sensitive, and some have changed name. You can try running

    <BOUT++ directory>/bin/bout-v5-input-file-upgrader.py delta_1/BOUT.inp

to automatically fix the most common issues. If these options above
are sometimes used depending on other options, you can call
`Options::setConditionallyUsed()` like:

    Options::root()["All"].setConditionallyUsed();

to mark a section or value as depending on other values, and so ignore
it in this check. Alternatively, if you're sure the above inputs are
not a mistake, you can set 'input:error_on_unused_options=false' to
turn off this check for unused options. You can always set
'input:validate=true' to check inputs without running the full
simulation.

Suggested alternatives:

Unused option 'All:bndry_all', did you mean:
        all:bndry_all

Unused option 'All:scale', did you mean:
        all:scale
        n:scale

Unused option 'NOUT', did you mean:
        nout
        solver:nout
        MXG
        MYG
        ZMAX
        ZMIN

Unused option 'TIMESTEP', did you mean:
        timestep

Unused option 'phiBoussinesq:flags', did you mean:
        phiBoussinesq:filter

Unused option 'phiSolver:type', did you mean:
        solver:type

```

Very happy to take suggestions on how to tweak the error message!

I've changed a few input options to be lowercase, but I realised things like `MXG` might be set in a grid file, and I don't want to deal with that kettle of worms.

Still need to fix most input files in tests and examples, but that needs the Python `BoutOptionsFile` fixes above.

Still need to implement `input:help`... that needs changes to printing `Options` I think.

**Comment by ZedThree** [comment]
Here's another fun issue: the available boundary regions may be different on each processor, so an input like `n:bndry_core` may be unused on one processor, while `n:bndry_target` may be used on another.

It might be possible to just mark everything named `bndry_*` to be ignored when checking unused options, but then we lose the ability to check if they are right. Maybe if the mesh registered possible regions before creating the specific ones on each processor? Tempted to leave that for another PR!

**Comment by johnomotani** [comment]
> Here's another fun issue: the available boundary regions may be different on each processor, so an input like n:bndry_core may be unused on one processor, while n:bndry_target may be used on another.

If fixing this had the side-effect of making all the boundary condition options always be printed to stdout, I wouldn't complain...

**Comment by ZedThree** [comment]
> If fixing this had the side-effect of making all the boundary condition options always be printed to stdout, I wouldn't complain...

I guess the most interesting thing would be to all-reduce the set of boundary regions to give the available options for this particular grid. Not sure exactly how one would do that, but there's almost certainly a solution.

**Comment by ZedThree** [comment]
I think this is nearing completion, it has turned out to be a lot more effort than I was anticipating! I've moved the check for unused options into its own function, `bout::checkForUnusedOptions()`. It's still called in `Solver::solve`, but now it can also be called in e.g. tests that don't use a time solver. Also means it will be much easier to move the check elsewhere -- I'm not even sure we need to wait till after calling `PhysicsModel::rhs`, `init` should be enough? That's for a future PR though!

I had to make the expression parse case sensitive as well, but I also improved its error message, so we get things like:

```
what():  Couldn't find generator 'Atan'. BOUT++ expressions are now case-sensitive, so you
may need to change your input file.
  1 + 0.1*Atan(y-3)*atan(4-y)
          ^~~~
           atan
  Did you mean 'atan'?
```

I've not written an auto-fixer for this, because I think it's likely much harder to do. The error message also only suggests the most likely candidate (if there is one), rather than all possible candidates, but it does point out where the error is at least.

This does mean there's a pretty good chance your input file will break!

---

@dschwoerer Please can you look at the Fedora build? Is it us or them?

```
warning: /var/cache/dnf/rawhide-2d95c80a1fa0a67d/packages/Lmod-8.4.15-2.fc34.x86_64.rpm: Header V4 RSA/SHA256 Signature, key ID 45719a39: NOKEY
Fedora - Rawhide - Developmental packages for t 1.6 MB/s | 1.6 kB     00:00    
GPG key at file:///etc/pki/rpm-gpg/RPM-GPG-KEY-fedora-35-x86_64 (0x9867C58F) is already installed
The GPG keys listed for the "Fedora - Rawhide - Developmental packages for the next Fedora release" repository are already installed but they are not correct for this package.
Check that the correct key URLs are configured for this repository.. Failing package is: Lmod-8.4.15-2.fc34.x86_64
 GPG Keys are configured as: file:///etc/pki/rpm-gpg/RPM-GPG-KEY-fedora-35-x86_64
<repeated lots for each package>
```

**Comment by dschwoerer** [comment]
> @dschwoerer <https://github.com/dschwoerer> Please can you look at the 
> Fedora build? Is it us or them?

Yes, that is a fedora issue. I guess the next release should have the 
fix, if it isn't fixed soon I will look into it.

**Comment by ZedThree** [comment]
It's possible that making options case-sensitive at the same time as erroring on unused options was a mistake. Too late now!

**Comment by bendudson** [comment]
@ZedThree Doing both in one PR has definitely complicated things, but as someone who's not had to do any actual work on it, I think it was the right choice: It gives a high confidence that people can't be caught out by the change to case sensitivity, and end up with input files that no longer do what they thought they did. Thanks for doing all this!

**Comment by ZedThree** [comment]
It has already caught a few bugs and mistakes that we wouldn't have noticed, that's for sure.

I've just discovered another fun thing: `time_report:show` is only used in `BoutFinalise`, so setting that will always trigger it as an unused option. So that will need to be used in tandem with `input:error_on_unused_options=false`. (Maybe that should be `input:no_error_on_unused_options` so it could work as a flag? `input:allow_unused_options`? Better names gladly received!)

Moving to a `BoutSim` class or something may bring a whole load of benefits -- we could make things like that a member and set the value in the constructor to be used later.

**Comment by ZedThree** [comment]
Fedora build is failing because it looks like they break the rolling version, rawhide, that we use every time there's a new release since at least Fedora 26. Something to do with not having the signing keys available until it's too late.

Everything else is passing now, so this is pretty much "ready". There's still the issue of boundary regions. In order to avoid all-to-all comms, each processor should work out the full set of possible boundary regions. `FieldData::setBoundary` can then iterate over the list of possible regions and mark them as conditionally used, before using the actually available regions. This way, the user would still get useful error messages if they specify a condition for a mistyped region or one that doesn't exist for the particular grid.

I think that might be best to wait for another PR so that we can refactor `BoutMesh` a little bit to reuse the boundary code to do this.

We should also completely remove `dump_format`, but I think that can also wait till we remove datafile stuff

File: examples/elm-pb/data/BOUT.inp
Comment (bendudson) [inline examples/elm-pb/data/BOUT.inp:52]: Should `shiftOutput` be changed to `shiftoutput` ? It looks like most CamelCase inputs have just become all lower case

**Review by johnomotani — COMMENTED** [review]
I've had a look through. I didn't have time to get my head around the fuzzy matching logic, so I've mostly checked the changes to input files.

File: src/sys/options.cxx
Comment (ZedThree) [reply src/sys/options.cxx:704]: This is covers the case where the `Options` is neither a value nor a section. I didn't quite pin down exactly when that can happen, but apparently it does. I'll try digging into it again or at least make the comment better!

**Comment by ZedThree** [comment]
Thanks for going through all the input files @johnomotani !

**Comment by ZedThree** [comment]
I've now got a solution for getting all the possible boundary names: copy the `BoutMesh` but change its rank. We can then do that at nine separate locations: at the outer edge to see if there's a SOL boundary, and then at each of the possible target locations (`y = 0, ny_inner - 1, ny_inner, ny - 1`) and each of the branch cuts (`y = jyseps*`) at `x = 0`. This neatly avoids all communication, which makes it easy to test, and is also O(1) so is just as fast on thousands of cores as it is on one core -- seems to take about 17ms in testing.

It needs #2245, so I'll wait till that's merged before adding it here

File: src/sys/options.cxx
Comment (ZedThree) [reply src/sys/options.cxx:704]: It turns out this can happen when calling `option.getSection()` with a section that doesn't exist. This calls `operator[]` which creates a new `Options`, with default values for `is_section` and `is_value` (both `false`).

A possible fix is:

```cpp
Options* Options::getSection(const std::string& name) {
  auto* section = &(*this)[name];
  section->is_section = true;
  return section;
}
```

but doing something like

```cpp
Options section = Options::root()["new section"];
```

still won't work.

Another solution would be to remove `is_section` and just have `is_value`. A new `Options` would default to being a section, but assignment would turn it into a value.

Having both let's us do things like:

```ini
name = value

[name]
stuff = foo
```

but do we really want to support having both a section and a value with the same name at the same level?

@bendudson Is there something else I'm missing that requires having both `is_section` and `is_value`?

Comment (bendudson) [reply src/sys/options.cxx:704]: I hoped to get rid of this too,  but at the moment we have a top level option `restart` and a section of the same name

**Comment by ZedThree** [comment]
Ah, I had a feeling there was one, but I couldn't remember it off the top
of my head and searching for an example seemed tricky! Might need to do so
to find any others

We could transform the command line flag into something like
"restart:restart = true" when converting it to an Option. And we have the
tooling to change it in input files too. There would be a discrepancy
between setting it on the command line and in an input file, which is a bit
unpalatable, but maybe not too much?

File: examples/boutcore/data/BOUT.inp
Comment (bendudson) [inline examples/boutcore/data/BOUT.inp:5]: Can these top-level options be removed? There are MXG, MYG settings in the `mesh` section too

File: tests/integrated/test-drift-instability/2fluid.cxx
Comment (bendudson) [reply tests/integrated/test-drift-instability/2fluid.cxx:145]: The main use for this is deciding whether the curvature should be modified: The `z` component changes between field aligned and shifted coordinates. Probably something that should be handled by coordinates/mesh/paralleltransform, so when you read a vector from the mesh, you get the components for the coordinate & parallel transform on that mesh.

**Review by bendudson — DISMISSED** [review]
Thanks @ZedThree ! This is quite a large PR. Probably good to merge soon before it conflicts with everything

**Comment by ZedThree** [comment]
I'd just like to implement the fix for the boundary condition checking before we merge this. Hopefully won't take too long!

Also, it's not necessary for this PR, but merging #2277 would also allow a small simplification here.

File: examples/boutcore/data/BOUT.inp
Comment (ZedThree) [reply examples/boutcore/data/BOUT.inp:5]: One for @dschwoerer ?

Comment (dschwoerer) [reply examples/boutcore/data/BOUT.inp:5]: If the top-level options are removed, the mesh section ones are used, and the simulation fails.

I'll push a fix.

Comment (ZedThree) [reply examples/boutcore/data/BOUT.inp]: Lovely, thanks @dschwoerer :+1:

**Comment by ZedThree** [comment]
Should've just merged this earlier, instead of trying to fix it all in this PR!

It looks like the failing tests are because `Mesh::getPossibleBoundaries` is much slower than I thought it was. I'm going to try a couple of things, but I might just revert the use of that to get this in.

**Comment by ZedThree** [comment]
It turns out the biggest cost of `getPossibleBoundaries` was in creating `Region`s on the copied meshes, which turns out to be quite expensive for production-sized meshes. Not constructing the regions reduces the cost of `getPossibleBoundaries` by 10^4!

**Comment by ZedThree** [comment]
Here's the changes from today:

https://github.com/boutproject/BOUT-dev/pull/2210/files/1901901823e382becdd2c19b389d43418766dd9f..df7879b3f1afd98b699b1de9c44f06ddb1467f57

**Review by johnomotani — APPROVED** [review]
Looks good to me!

**Pull request merged**
